File: labview/src/auth/session.ts
/**
 * Sessions: a signed token in a cookie, and nothing on the server.
 *
 * The token carries who signed in, how, when it was issued and when it stops being
 * valid, MAC'd with a secret. There is no session store, which is deliberate — a
 * dashboard that degrades to "sign in again" after a restart is a better trade than a
 * database, and it means two replicas behind the same proxy work with no shared state
 * beyond `auth.session.secret`.
 *
 * The one piece of server state is the revocation set, so that signing out
 * invalidates the token rather than merely dropping the browser's copy of it. It is
 * in memory and bounded, and a restart clears it along with every session it could
 * apply to — consistent, if not durable.
 *
 * The clock is a parameter everywhere. Every rule here — expiry, the transient
 * cookie's five-minute window, pruning — is a comparison against `now`, and the only
 * way to assert those is to be able to move it.
 *
 * Cookie handling is written out rather than delegated to `@fastify/cookie`, for the
 * reason `cookiePairs` in `enrich/http.ts` gives: this is `split("; ")` and a header
 * string, and a dependency for it would have to be audited, updated and shipped.
 */
import { createHmac, randomBytes, timingSafeEqual } from "node:crypto";
import type { LoginMethod } from "../model/types.js";
import { isValidUsername } from "../model/access.js";

/** Version prefix, so a format change is a rejected token rather than a misread one. */
const VERSION = "v1";

export interface SessionPayload {
  /** Username. Always one {@link isValidUsername} accepts. */
  u: string;
  /** Which method signed this session in. */
  via: LoginMethod;
  /** Issued at, epoch seconds. */
  iat: number;
  /** Expires at, epoch seconds. */
  exp: number;
  /** Token id, for revocation on sign-out. */
  jti: string;
}

export type SessionRejection = "malformed" | "signature" | "expired" | "revoked";

export type SessionCheck =
  | { ok: true; payload: SessionPayload }
  | { ok: false; reason: SessionRejection };

function b64url(buf: Buffer): string {
  return buf.toString("base64url");
}

function mac(data: string, secret: string): string {
  return b64url(createHmac("sha256", secret).update(data).digest());
}

/**
 * Constant-time string compare that tolerates a length difference.
 *
 * `timingSafeEqual` throws on mismatched lengths, and the obvious `a.length ===
 * b.length &&` guard in front of it leaks the length — which for a fixed-width MAC is
 * nothing, but this helper is also used on the OIDC `state`. Hashing both sides first
 * makes the comparison fixed-width regardless of the inputs.
 */
export function safeEqual(a: string, b: string): boolean {
  const ha = createHmac("sha256", "compare").update(a).digest();
  const hb = createHmac("sha256", "compare").update(b).digest();
  return timingSafeEqual(ha, hb);
}

/**
 * `v1.<payload>.<mac>` — a signed, *not encrypted*, JSON payload.
 *
 * Both segments are base64url, so the result is safe in a cookie value with no
 * escaping: no `;`, no `,`, no whitespace, no `=` padding.
 *
 * Not encrypted because nothing in either payload is a secret — a username the
 * browser's owner already knows, and an expiry. What matters is that neither can be
 * *changed*, which is what the MAC is for.
 */
export function signPayload(payload: unknown, secret: string): string {
  const body = b64url(Buffer.from(JSON.stringify(payload), "utf8"));
  return `${VERSION}.${body}.${mac(body, secret)}`;
}

/**
 * The payload of a token whose MAC verifies, or `undefined`.
 *
 * Checks only the signature: what the fields have to contain is the caller's
 * question, and keeping the two apart is what lets the session token and the OIDC
 * transient cookie share this without sharing a shape.
 */
export function unsignPayload(token: string, secret: string): unknown | undefined {
  const parts = token.split(".");
  if (parts.length !== 3) return undefined;
  const [version, body, signature] = parts;
  if (version !== VERSION || !body || !signature) return undefined;
  if (!safeEqual(signature, mac(body, secret))) return undefined;
  try {
    return JSON.parse(Buffer.from(body, "base64url").toString("utf8")) as unknown;
  } catch {
    return undefined;
  }
}

/** A fresh session token for `user`, valid for `ttlMinutes` from `now`. */
export function issueSession(
  user: string,
  via: LoginMethod,
  secret: string,
  ttlMinutes: number,
  now: Date,
): { token: string; payload: SessionPayload } {
  const iat = Math.floor(now.getTime() / 1000);
  const payload: SessionPayload = {
    u: user,
    via,
    iat,
    exp: iat + Math.max(1, Math.floor(ttlMinutes)) * 60,
    jti: randomBytes(12).toString("base64url"),
  };
  return { token: signPayload(payload, secret), payload };
}

/**
 * Check a token, in the order the checks cost: shape, then signature, then expiry,
 * then revocation.
 *
 * Expiry after the signature on purpose — an unsigned token's claimed `exp` is not
 * worth reading, and reporting `expired` for one would tell a forger their guess had
 * been parsed.
 */
export function verifySession(
  token: string,
  secret: string,
  now: Date,
  revoked?: { has(jti: string): boolean },
): SessionCheck {
  const raw = unsignPayload(token, secret);
  if (raw === undefined) {
    // A malformed token and a bad signature are one answer to the caller; they are
    // distinguished here only so a log line can say which, and `unsignPayload`
    // deliberately does not tell them apart.
    return { ok: false, reason: token.split(".").length === 3 ? "signature" : "malformed" };
  }
  const p = raw as Partial<SessionPayload>;
  if (
    typeof p.u !== "string" ||
    !isValidUsername(p.u) ||
    (p.via !== "passwd" && p.via !== "oidc") ||
    typeof p.iat !== "number" ||
    typeof p.exp !== "number" ||
    typeof p.jti !== "string" ||
    !p.jti
  ) {
    return { ok: false, reason: "malformed" };
  }
  if (p.exp * 1000 <= now.getTime()) return { ok: false, reason: "expired" };
  if (revoked?.has(p.jti)) return { ok: false, reason: "revoked" };
  return { ok: true, payload: p as SessionPayload };
}

/**
 * Signed-out token ids, until they would have expired anyway.
 *
 * Bounded twice over: entries are dropped once `now` passes their own expiry, and the
 * map is capped so that a script calling `/api/logout` in a loop cannot grow it
 * without limit. At the cap the oldest expiry goes first, which is the entry closest
 * to being useless.
 */
export class SessionRevocations {
  private readonly until = new Map<string, number>();

  constructor(private readonly max = 10_000) {}

  revoke(jti: string, expEpochSeconds: number, now: Date): void {
    this.prune(now);
    if (this.until.size >= this.max) {
      let oldest: string | undefined;
      let oldestExp = Infinity;
      for (const [k, v] of this.until) {
        if (v < oldestExp) {
          oldest = k;
          oldestExp = v;
        }
      }
      if (oldest !== undefined) this.until.delete(oldest);
    }
    this.until.set(jti, expEpochSeconds);
  }

  has(jti: string): boolean {
    return this.until.has(jti);
  }

  prune(now: Date): void {
    const cutoff = Math.floor(now.getTime() / 1000);
    for (const [k, v] of this.until) if (v <= cutoff) this.until.delete(k);
  }

  get size(): number {
    return this.until.size;
  }
}

export interface CookieOptions {
  name: string;
  value: string;
  /** Seconds. `0` expires the cookie immediately, which is how sign-out clears it. */
  maxAgeSeconds: number;
  secure: boolean;
  path: string;
}

/**
 * One `Set-Cookie` value.
 *
 * `HttpOnly` because no script has any reason to read a session token, and it is the
 * difference between an XSS bug being a bug and being an account takeover.
 *
 * `SameSite=Lax` rather than `Strict`: `Strict` withholds the cookie on a
 * cross-site *navigation*, which is exactly what returning from an identity provider
 * is — the OIDC callback would arrive with no session and loop. `Lax` sends it on a
 * top-level GET and withholds it on a cross-site POST, which is the CSRF case that
 * matters, and the POST routes check `Origin` as well.
 */
export function serializeCookie(o: CookieOptions): string {
  const parts = [
    `${o.name}=${o.value}`,
    `Path=${o.path}`,
    `Max-Age=${Math.max(0, Math.floor(o.maxAgeSeconds))}`,
    "HttpOnly",
    "SameSite=Lax",
  ];
  if (o.secure) parts.push("Secure");
  return parts.join("; ");
}

/** The value of one cookie from a request's `Cookie` header. First wins. */
export function readCookie(header: string | undefined, name: string): string | undefined {
  if (!header) return undefined;
  for (const token of header.split(";")) {
    const eq = token.indexOf("=");
    if (eq <= 0) continue;
    if (token.slice(0, eq).trim() !== name) continue;
    return token.slice(eq + 1).trim();
  }
  return undefined;
}

/**
 * The scheme the *browser* used, which is the only one that decides whether a
 * `Secure` cookie will come back.
 *
 * LabView is meant to sit behind a reverse proxy, so its own socket is very often
 * plain HTTP while the request that reached it was HTTPS. Reading
 * `X-Forwarded-Proto` is the standard way to know, and the first value is the
 * client-facing hop when several proxies have appended to it.
 */
export function effectiveProtocol(forwardedProto: string | undefined, ownProtocol: string): string {
  const first = forwardedProto?.split(",")[0]?.trim().toLowerCase();
  return first || ownProtocol.toLowerCase();
}

/**
 * The host the browser addressed, for the `Origin` check and for a derived redirect URI.
 *
 * Same reasoning as {@link effectiveProtocol}: behind a proxy the request's own `Host`
 * may be an internal name, and `X-Forwarded-Host` is what the browser typed. Only the
 * host is compared anywhere — never the scheme — because a TLS-terminating proxy makes
 * the two disagree by design.
 */
export function effectiveHost(forwardedHost: string | undefined, ownHost: string): string {
  const first = forwardedHost?.split(",")[0]?.trim().toLowerCase();
  return first || ownHost.toLowerCase();
}

/**
 * Whether an `Origin` header belongs to the same host the request arrived on.
 *
 * The CSRF check for every POST while enforcing, on top of `SameSite=Lax`. A **missing**
 * `Origin` passes: browsers send it on every cross-site POST, so its absence means the
 * request did not come from a page — `curl`, a script, a health checker — and those have
 * no ambient cookie to abuse. An `Origin` that is present and different is refused.
 */
export function originAllowed(origin: string | undefined, host: string): boolean {
  if (!origin) return true;
  try {
    return new URL(origin).host.toLowerCase() === host.toLowerCase();
  } catch {
    // `Origin: null` — a sandboxed iframe or a redirected form post. Not this host.
    return false;
  }
}

/**
 * Whether to mark cookies `Secure`.
 *
 * `auto` follows the effective scheme, which is right in both deployments that
 * matter: behind a TLS proxy the cookie is `Secure`, and on a plain-HTTP LAN address
 * it is not — because a `Secure` cookie over HTTP is never stored, and the symptom is
 * a login form that accepts the password and comes straight back, with nothing in any
 * log to say why. `true` and `false` exist for an operator whose proxy does not set
 * the header.
 */
export function shouldSecureCookie(setting: string, protocol: string): boolean {
  if (setting === "true") return true;
  if (setting === "false") return false;
  return protocol === "https";
}

File: labview/internal/access/session.go
package access

import (
	"crypto/hmac"
	"crypto/rand"
	"crypto/sha256"
	"encoding/base64"
	"encoding/hex"
	"encoding/json"
	"errors"
	"net/http"
	"sort"
	"strings"
	"sync"
	"time"

	"github.com/nrosier/labview/internal/payload"
)

// Sessions are signed, not stored (§19).
//
// A token carries its own claims and a MAC over them, so there is no session table to grow, to
// synchronise or to lose on restart. **No persistence across restarts is a stated non-goal** — a new
// signing secret at startup invalidates every outstanding token, which for a lab viewer is the right
// trade: the cost is one sign-in and the benefit is no state.
//
// The whole format is one line:
//
//	v1.<base64url(payload)>.<base64url(HMAC-SHA256(payload))>
//
// Version first, because a format change has to be distinguishable from a corrupt token. base64url
// unpadded, because the value goes in a cookie and `=` and `/` are both awkward there.

// The session token's shape.
const (
	// TokenVersion is the prefix every token carries.
	TokenVersion = "v1"

	// MaxTokenBytes bounds what will be parsed. A token is around 200 bytes; anything approaching a
	// kilobyte is not a token, and refusing it before base64-decoding means a megabyte cookie costs
	// nothing.
	MaxTokenBytes = 1024
)

// Claims is the signed payload: `{u, via, iat, exp, jti}` (§19).
//
// The names are short because they are transmitted on every request. The fields are exactly five —
// there is no room here for a role, a permission or a scope, since §19 states no roles as a non-goal
// and a claim nobody checks is a claim somebody will eventually trust.
type Claims struct {
	// U is the username, always already sanitised (§16). Sanitised before signing rather than after
	// parsing, so a hostile name cannot be inside a *valid* token at all.
	U string `json:"u"`

	// Via is which method minted this session, for the session endpoint and the log.
	Via payload.LoginMethod `json:"via"`

	// Iat and Exp are Unix seconds. Seconds rather than milliseconds because a cookie's Max-Age is in
	// seconds and two units for one lifetime is two chances to be off by a thousand.
	Iat int64 `json:"iat"`
	Exp int64 `json:"exp"`

	// Jti is the identifier logout revokes. Random per session and never derived from the username or
	// the time: a predictable identifier would let anybody revoke anybody's session.
	Jti string `json:"jti"`
}

// User is what this session says about who is signed in.
func (c Claims) User() payload.SessionUser {
	return payload.SessionUser{Name: Username(c.U), Via: c.Via}
}

// Expiry is Exp as a time.
func (c Claims) Expiry() time.Time { return time.Unix(c.Exp, 0) }

// ErrNoSession is what a request with no cookie at all produces. It is separate from a rejection
// because it is not one: an anonymous request to an enforcing server is the ordinary case, and
// reporting it as `malformed` would fill the log with every unauthenticated hit.
var ErrNoSession = errors.New("no session cookie")

// Signer mints and verifies tokens.
//
// The secret is held here and nowhere else, and there is no accessor for it: a struct with no getter
// cannot have its key logged by a handler that means well (I6).
type Signer struct {
	secret []byte
	ttl    time.Duration

	// revoked is the one revocation set (§19). It is on the signer rather than beside it because
	// verification consults it and there is exactly one of both.
	revoked *revocations
}

// SessionTTLFloor is the shortest session worth issuing. A TTL of zero or less would mint a token
// that has already expired, so it is treated as *use the default* rather than as a lock-out (I4).
const SessionTTLFloor = time.Minute

// DefaultSessionTTL matches §3's `auth.session.ttlMinutes` default of 720.
const DefaultSessionTTL = 720 * time.Minute

// NewSigner builds a signer. A secret shorter than 32 bytes is stretched by hashing rather than
// refused: refusing would be a lock-out over a configuration value, and hashing a weak secret does
// not make it strong but does make it the right length for HMAC.
func NewSigner(secret string, ttl time.Duration) *Signer {
	key := []byte(secret)
	if len(key) < sha256.Size {
		sum := sha256.Sum256(key)
		key = sum[:]
	}
	if ttl < SessionTTLFloor {
		ttl = DefaultSessionTTL
	}
	return &Signer{secret: key, ttl: ttl, revoked: &revocations{}}
}

// RandomSecret is the secret used when configuration names none.
//
// A random secret per process rather than a constant: a committed default signing key means anybody
// with the source can mint a session for any LabView that never set one, which is every default
// install. The cost is that sessions do not survive a restart, which §19 already accepts.
func RandomSecret() string {
	b := make([]byte, 32)
	if _, err := rand.Read(b); err != nil {
		// Unreachable on any supported platform. A degraded secret still gates; a refusal to start
		// does not (I4).
		return "labview session secret fallback"
	}
	return hex.EncodeToString(b)
}

// TTL is how long a minted session lasts.
func (s *Signer) TTL() time.Duration { return s.ttl }

// Mint produces a token for user signed in via method, valid from now.
//
// now is a parameter because §19 requires every pure part to take the current time — and because a
// test for *expired* that has to wait for a real clock is a test nobody runs.
func (s *Signer) Mint(user string, via payload.LoginMethod, now time.Time) (string, Claims, error) {
	jti, err := newJTI()
	if err != nil {
		return "", Claims{}, err
	}

	claims := Claims{
		// Sanitised here, at the moment of minting. Every later reader of a valid token therefore
		// holds a name that already satisfies the pattern (§16).
		U:   Username(user),
		Via: via,
		Iat: now.Unix(),
		Exp: now.Add(s.ttl).Unix(),
		Jti: jti,
	}
	body, err := json.Marshal(claims)
	if err != nil {
		return "", Claims{}, err
	}

	encoded := enc(body)
	return TokenVersion + "." + encoded + "." + enc(s.mac([]byte(encoded))), claims, nil
}

// Verify checks a token and returns its claims.
//
// **The checks run in exactly §19's order: shape, MAC, expiry, revocation.** The order is the
// security property, not a style: checking expiry before the MAC would answer a question about a
// document that has not been shown to be ours, and reporting `expired` for a forged token would tell
// a forger that their signature was the only thing wrong. Revocation last because it is the only
// check that touches shared state, and an unauthenticated caller should not be able to make the
// server take a lock.
func (s *Signer) Verify(token string, now time.Time) (Claims, payload.SessionRejection, error) {
	if len(token) > MaxTokenBytes {
		return Claims{}, payload.RejectMalformed, errors.New("session token is too long")
	}

	// Shape.
	version, body, mac, ok := cut(token)
	if !ok || version != TokenVersion {
		return Claims{}, payload.RejectMalformed, errors.New("session token is not " + TokenVersion + ".payload.signature")
	}
	given, err := dec(mac)
	if err != nil {
		return Claims{}, payload.RejectMalformed, errors.New("session signature is not base64url")
	}

	// MAC — before the payload is unmarshalled, so no attacker-chosen structure is interpreted until
	// it has been shown to be ours.
	if !equalHashed(given, s.mac([]byte(body))) {
		return Claims{}, payload.RejectSignature, errors.New("session signature does not verify")
	}

	raw, err := dec(body)
	if err != nil {
		// Reachable only for a token we signed ourselves, so it is a bug rather than an attack — but
		// it is still reported rather than trusted.
		return Claims{}, payload.RejectMalformed, errors.New("session payload is not base64url")
	}
	var claims Claims
	if err := json.Unmarshal(raw, &claims); err != nil {
		return Claims{}, payload.RejectMalformed, errors.New("session payload is not the expected object")
	}
	if claims.U == "" || claims.Jti == "" || claims.Exp == 0 {
		return Claims{}, payload.RejectMalformed, errors.New("session payload is missing a required claim")
	}

	// Expiry.
	if !now.Before(claims.Expiry()) {
		return claims, payload.RejectExpired, errors.New("session expired")
	}

	// Revocation.
	if s.revoked.has(claims.Jti) {
		return claims, payload.RejectRevoked, errors.New("session was signed out")
	}

	return claims, "", nil
}

// Revoke adds a session's identifier to the revocation set (§19).
func (s *Signer) Revoke(c Claims, now time.Time) { s.revoked.add(c.Jti, c.Expiry(), now) }

// Revocations is how many identifiers are held, for a test and a diagnostic.
func (s *Signer) Revocations() int { return s.revoked.len() }

// mac is the HMAC over the encoded payload.
//
// Over the *encoded* payload rather than the raw JSON, because that is the byte string actually
// present in the token: signing the decoded form would mean two different tokens could carry the same
// signature if base64 had any slack, and unpadded base64url of an attacker-chosen string is exactly
// where that slack lives.
func (s *Signer) mac(body []byte) []byte {
	m := hmac.New(sha256.New, s.secret)
	m.Write(body)
	return m.Sum(nil)
}

// equalHashed compares two MACs in constant time, **hashing both sides first** (§19).
//
// hmac.Equal is already constant-time in its content, but it returns immediately when the lengths
// differ — so a caller could learn the MAC's length by timing, and more usefully could tell
// *truncated* from *wrong*. Hashing makes both sides 32 bytes whatever came in, so length tells
// nothing and there is exactly one comparison path.
func equalHashed(a, b []byte) bool {
	ha, hb := sha256.Sum256(a), sha256.Sum256(b)
	return hmac.Equal(ha[:], hb[:])
}

// cut splits a token into its three parts without allocating a slice.
func cut(token string) (version, body, mac string, ok bool) {
	first := strings.Index(token, ".")
	if first < 0 {
		return "", "", "", false
	}
	rest := token[first+1:]
	second := strings.Index(rest, ".")
	if second < 0 {
		return "", "", "", false
	}
	body, mac = rest[:second], rest[second+1:]
	if body == "" || mac == "" || strings.Contains(mac, ".") {
		return "", "", "", false
	}
	return token[:first], body, mac, true
}

func enc(b []byte) string          { return base64.RawURLEncoding.EncodeToString(b) }
func dec(s string) ([]byte, error) { return base64.RawURLEncoding.DecodeString(s) }

// newJTI produces a session identifier from 16 random bytes.
func newJTI() (string, error) {
	b := make([]byte, 16)
	if _, err := rand.Read(b); err != nil {
		return "", err
	}
	return enc(b), nil
}

// ---------------------------------------------------------------------------
// Revocation
// ---------------------------------------------------------------------------

// MaxRevocations is the cap §19 states.
//
// A bound is needed because logout is reachable by anybody holding a token, and a set that grew
// without limit would be a memory leak with a public trigger. 10 000 is more sign-outs than a lab
// produces in a session lifetime.
const MaxRevocations = 10000

// revocations is the one revocation set: identifier to the expiry after which it no longer matters.
//
// Keyed on the identifier and valued on the expiry, because that is what makes pruning possible: an
// entry whose session has expired is already refused by the expiry check, so keeping it would be
// holding state to answer a question that is answered earlier.
type revocations struct {
	mu sync.Mutex
	at map[string]time.Time
}

// add revokes an identifier, **pruning by expiry on every write** (§19).
//
// Pruning on write rather than on a timer: writes are rare (a logout), reads are constant (every
// request), and a background goroutine to sweep a map this small would be more moving parts than the
// map.
func (r *revocations) add(jti string, expiry, now time.Time) {
	if jti == "" {
		return
	}

	r.mu.Lock()
	defer r.mu.Unlock()

	if r.at == nil {
		r.at = map[string]time.Time{}
	}
	for held, exp := range r.at {
		if !now.Before(exp) {
			delete(r.at, held)
		}
	}
	r.at[jti] = expiry

	if len(r.at) > MaxRevocations {
		r.evict()
	}
}

// evict drops down to the cap, **earliest expiry first** (§19). The caller holds the lock.
//
// Earliest expiry rather than oldest insertion, because an entry that expires soonest is the one whose
// removal costs least: it is about to stop mattering anyway. Evicting by insertion order would drop a
// long-lived session's revocation while keeping one that expires in a minute — which is exactly
// backwards, since the dropped one is the token still usable.
func (r *revocations) evict() {
	type entry struct {
		jti string
		exp time.Time
	}
	all := make([]entry, 0, len(r.at))
	for jti, exp := range r.at {
		all = append(all, entry{jti, exp})
	}
	sort.Slice(all, func(i, j int) bool {
		if all[i].exp.Equal(all[j].exp) {
			// Tie broken on the identifier so eviction is deterministic (I7) rather than dependent on
			// map order.
			return all[i].jti < all[j].jti
		}
		return all[i].exp.Before(all[j].exp)
	})
	for i := 0; i < len(all) && len(r.at) > MaxRevocations; i++ {
		delete(r.at, all[i].jti)
	}
}

func (r *revocations) has(jti string) bool {
	r.mu.Lock()
	defer r.mu.Unlock()
	_, ok := r.at[jti]
	return ok
}

func (r *revocations) len() int {
	r.mu.Lock()
	defer r.mu.Unlock()
	return len(r.at)
}

// ---------------------------------------------------------------------------
// The cookie
// ---------------------------------------------------------------------------

// DefaultCookieName is the session cookie's name when configuration names none.
const DefaultCookieName = "labview_session"

// Cookies mints and clears the session cookie.
type Cookies struct {
	// Name is the cookie's name. Empty takes DefaultCookieName.
	Name string

	// Secure is `auto`, `true` or `false` (§3). `auto` follows the effective scheme.
	Secure string
}

func (c Cookies) name() string {
	if strings.TrimSpace(c.Name) == "" {
		return DefaultCookieName
	}
	return c.Name
}

// Set builds the Set-Cookie for a minted token.
//
// The four properties are not configurable, because each of them is the thing that makes the cookie
// safe rather than a preference:
//
//   - HttpOnly, so a script cannot read the token. LabView renders no untrusted HTML, but a session
//     readable by JavaScript is one XSS away from being stolen rather than merely being used.
//   - SameSite=Lax, which is half of the CSRF defence (§19) — the other half is the Origin check.
//   - Path=/, because the UI, the API and the OIDC routes are all under it.
//   - Max-Age from the TTL, so a browser drops the cookie at the same moment the server stops
//     honouring it, and a user is not shown a signed-in shell that 401s on every request.
func (c Cookies) Set(r *http.Request, token string, ttl time.Duration) *http.Cookie {
	return &http.Cookie{
		Name:     c.name(),
		Value:    token,
		Path:     "/",
		HttpOnly: true,
		SameSite: http.SameSiteLaxMode,
		Secure:   c.secure(r),
		MaxAge:   int(ttl.Seconds()),
	}
}

// Clear builds the Set-Cookie that removes the session.
//
// Same name, path and flags as Set. A clearing cookie whose attributes differ from the one it is
// clearing is not the same cookie as far as a browser is concerned, and the session would survive the
// logout it was told about.
func (c Cookies) Clear(r *http.Request) *http.Cookie {
	return &http.Cookie{
		Name:     c.name(),
		Value:    "",
		Path:     "/",
		HttpOnly: true,
		SameSite: http.SameSiteLaxMode,
		Secure:   c.secure(r),
		MaxAge:   -1,
	}
}

// Token reads the session token from a request, or ErrNoSession.
func (c Cookies) Token(r *http.Request) (string, error) {
	cookie, err := r.Cookie(c.name())
	if err != nil || cookie.Value == "" {
		return "", ErrNoSession
	}
	return cookie.Value, nil
}

// secure resolves the `Secure` attribute.
func (c Cookies) secure(r *http.Request) bool {
	switch c.Secure {
	case "true":
		return true
	case "false":
		return false
	default:
		return EffectiveScheme(r) == "https"
	}
}

// EffectiveScheme is the scheme the *client* used, which is not always the one this server saw.
//
// **`X-Forwarded-Proto` first** (§19). LabView normally sits behind the very reverse proxy it
// documents, terminating TLS and forwarding plain HTTP — so `r.TLS == nil` is the usual case for a
// server whose users are on https, and trusting it would mean never setting Secure in the one
// deployment that most needs it.
//
// The trade is stated plainly: a header is client-controlled unless a proxy overwrites it, so a
// direct-to-LabView request can claim https and get a Secure cookie it cannot then send back. That
// fails closed — the user cannot sign in over plain HTTP — whereas the other mistake ships a session
// cookie in the clear. Only the first value of a comma-joined list is read, since that is the
// original client's.
func EffectiveScheme(r *http.Request) string {
	if r == nil {
		return "http"
	}
	if forwarded := r.Header.Get("X-Forwarded-Proto"); forwarded != "" {
		first := forwarded
		if i := strings.Index(first, ","); i >= 0 {
			first = first[:i]
		}
		if scheme := strings.ToLower(strings.TrimSpace(first)); scheme == "https" || scheme == "http" {
			return scheme
		}
	}
	if r.TLS != nil {
		return "https"
	}
	return "http"
}
